shared/pingpong/accounts.go

```diff
@@ -186,133 +194,156 @@ func prepareAssets(accounts map[string]uint64, client libgoal.Client, cfg PpConf
+				return
+			}
+			if len(account.Assets) >= expectedAssets {
+				break
 			}
+			if time.Now().After(start.Add(3 * time.Minute)) {
+				err = fmt.Errorf("asset opting in took too long")
+				fmt.Printf("Error: %s\n", err.Error())
+				return
+			}
+			time.Sleep(time.Second)
 		}
 	}
-	time.Sleep(time.Second * 15)
 
-	// Step 3) Evenly distribute the assets across all participant accounts
-	for addr := range accounts {
+	// Step 3) Evenly distribute the assets across all opted-in accounts
+	for k, creator := range allAssets {
 		if !cfg.Quiet {
-			fmt.Printf("Distributing assets from account %v\n", addr)
+			fmt.Printf("Distributing asset %+v from account %v\n", k, creator)
 		}
-		addrAccount, addrErr := client.AccountInformation(addr)
-		if addrErr != nil {
+		creatorAccount, creatorErr := client.AccountInformation(creator)
+		if creatorErr != nil {
 			fmt.Printf("Cannot lookup source account\n")
-			err = addrErr
+			err = creatorErr
 			return
 		}
+		assetParams := creatorAccount.AssetParams
 
-		assetParams := addrAccount.AssetParams
-		if !cfg.Quiet {
-			fmt.Printf("Distributing  %d assets\n", len(assetParams))
-		}
-
-		// Distribute assets to each account
-		for k := range assetParams {
+		for _, addr := range optIns[k] {
+			assetAmt := assetParams[k].Total / uint64(len(optIns[k]))
 			if !cfg.Quiet {
-				fmt.Printf("Distributing asset %v \n", k)
+				fmt.Printf("Distributing assets from %v to %v \n", creator, addr)
 			}
 
-			assetAmt := assetParams[k].Total / uint64(len(accounts))
-			for addr2 := range accounts {
-				if addr != addr2 {
-					if !cfg.Quiet {
-						fmt.Printf("Distributing assets from %v to %v \n", addr, addr2)
-					}
-
-					tx, sendErr := constructTxn(addr, addr2, cfg.MaxFee, assetAmt, k, CreatablesInfo{}, client, cfg)
-					if sendErr != nil {
-						fmt.Printf("Cannot transfer asset %v from account %v\n", k, addr)
-						err = sendErr
-						return
-					}
-
-					_, err = signAndBroadcastTransaction(accounts, addr, tx, client, cfg)
-					if err != nil {
-						_, _ = fmt.Fprintf(os.Stderr, "signing and broadcasting asset distribution failed with error %v\n", err)
-						return
-					}
+			tx, sendErr := constructTxn(creator, addr, cfg.MaxFee, assetAmt, k, CreatablesInfo{}, client, cfg)
+			if sendErr != nil {
+				fmt.Printf("Cannot transfer asset %v from account %v\n", k, creator)
+				err = sendErr
+				return
+			}
 
-					totalSent++
-					throttleTransactionRate(startTime, cfg, totalSent)
-				}
+			_, err = signAndBroadcastTransaction(accounts, creator, tx, client, cfg)
+			if err != nil {
+				_, _ = fmt.Fprintf(os.Stderr, "signing and broadcasting asset distribution failed with error %v\n", err)
+				return
 			}
-			// append the asset to the result assets
-			resultAssetMaps[k] = assetParams[k]
+
+			totalSent++
+			throttleTransactionRate(startTime, cfg, totalSent)
 		}
+		// append the asset to the result assets
+		resultAssetMaps[k] = assetParams[k]
 	}
 	time.Sleep(time.Second * 10)
```

Comment:
I'm not sure what's the purpose of the 10 second sleep. If we want to make sure that all our transactions were included in a block(s), we should test that explicitly.

Reply:
yeah, I think that was the intention.

Reply:
done